chain-tx-enclave-next/mls/src/astree.rs

```diff
@@ -0,0 +1,185 @@
+use crate::ciphersuite::{CipherSuite, HkdfExt};
+use crate::tree_math::{LeafSize, NodeSize};
+use hkdf::Hkdf;
+use secrecy::{ExposeSecret, SecretVec};
+use sha2::digest::{generic_array, BlockInput, FixedOutput, Input, Reset};
+use std::marker::PhantomData;
+
+/// "Application Secret Tree" in specs
+/// treeBaseKeySource in ref impl
+struct TreeBaseKeySource<D: Input + BlockInput + FixedOutput + Reset + Default + Clone> {
+    pub secrets: Vec<Option<SecretVec<u8>>>,
+    pub marker: PhantomData<D>,
+}
+
+impl<D: Input + BlockInput + FixedOutput + Reset + Default + Clone> TreeBaseKeySource<D> {
+    /// astree_node_[root]_secret = application_secret
+    pub fn new(application_secret: SecretVec<u8>, leaf_count: LeafSize) -> Self {
+        let root_index = NodeSize::root(leaf_count);
+        let node_len = NodeSize::node_width(leaf_count);
+        let mut secrets = Vec::with_capacity(node_len.0);
+        for _ in 0..node_len.0 {
+            secrets.push(None);
+        }
+        secrets[root_index.0] = Some(application_secret);
+        Self {
+            secrets,
+            marker: PhantomData,
+        }
+    }
+
+    pub fn get_base_secret(
+        &mut self,
+        group_context_hash: Vec<u8>,
```

Comment:
is it possible with
&[u8]?

Reply:
the context hash is usually put in this label structure: https://github.com/crypto-com/chain/blob/master/chain-tx-enclave-next/mls/src/ciphersuite.rs#L27
that's needed in these functions: https://github.com/mlswg/mls-protocol/blob/master/draft-ietf-mls-protocol.md#key-schedule

so, in the current code, if I put `&[u8]`, it'll need to call `.to_vec` down the line anyway... 

It's probably possible to remove some of this hash-vec cloning, but unless there's a simple and obvious change, it may be better to delay that until the code is more complete